### PR TITLE
Go: generate equality methods

### PR DESCRIPTION
**Pull request by K-Phoen**
Implements #551

Still to-do:

- [x] panels equality? Or is `reflect.DeepEqual` enough? → seems to be enough
- [x] tests → basic coverage for now

### CODE REVIEW

File: internal/jennies/golang/templates/types/struct_equality_method.tmpl [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	{{- if .Type.IsAny }}
+		{{- $dereference := ternary "*" "" .Dereference }}
+		// is DeepEqual good enough here?
+		if !reflect.DeepEqual({{ $dereference }}{{ .SelfName }}, {{ $dereference }}{{ .OtherName }}) {
+			return false
+		}
+	{{- else if resolvesToArray .Type }}
+		{{- $needsDereference := and .Type.IsRef .Type.Nullable }}
+		{{- $dereference := ternary "*" "" $needsDereference }}
+		{{- $depth := default 1  .Depth }}
```
Comment (spinillos) [inline internal/jennies/golang/templates/types/struct_equality_method.tmpl:18]:
```suggestion
		{{- $depth := default 1 .Depth }}
```

File: internal/jennies/golang/templates/types/struct_equality_method.tmpl [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+		for i{{ $depth }} := range {{ $dereference }}{{ .SelfName }} {
+		{{- $selfKey := print (ternary (print "(*" .SelfName ")") .SelfName $needsDereference) "[i" $depth "]" }}
+		{{- $otherKey := print (ternary (print "(*" .OtherName ")") .OtherName $needsDereference) "[i" $depth "]" }}
+			{{- template "type_equality_check" (dict "Type" (resolveRefs .Type).Array.ValueType "Nullable" (resolveRefs .Type).Array.ValueType.Nullable "Dereference" false "SelfName" $selfKey "OtherName" $otherKey "Depth" (add1 $depth)) }}
+		}
+		{{- if $needsDereference }}}{{ end }}
+	{{- else if resolvesToMap .Type }}
+		{{- $needsDereference := and .Type.IsRef .Type.Nullable }}
+		{{- $dereference := ternary "*" "" $needsDereference }}
+		{{- $depth := default 1  .Depth }}
```
Comment (spinillos) [inline internal/jennies/golang/templates/types/struct_equality_method.tmpl:40]:
```suggestion
		{{- $depth := default 1 .Depth }}
```

File: internal/jennies/golang/templates/types/struct_equality_method.tmpl [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+		// is DeepEqual good enough here?
+		if !reflect.DeepEqual({{ $dereference }}{{ .SelfName }}, {{ $dereference }}{{ .OtherName }}) {
+			return false
+		}
+	{{- else if resolvesToArray .Type }}
+		{{- $needsDereference := and .Type.IsRef .Type.Nullable }}
+		{{- $dereference := ternary "*" "" $needsDereference }}
+		{{- $depth := default 1  .Depth }}
+		{{- if $needsDereference }}
+			if !(({{ .SelfName }} == nil && {{ .OtherName }} == nil) || ({{ .SelfName }} != nil && {{ .OtherName }} != nil)) {
```
Comment (spinillos) [inline internal/jennies/golang/templates/types/struct_equality_method.tmpl:20]: Isn't the same than: 
```
{{ .SelfName }} == nil && {{ .OtherName }} != nil || {{ .SelfName }} != nil && {{ .OtherName }} == nil
```

You can remove some parentheses. And same for the one in map.

File: Makefile [hunk 4]
```diff
@@ -48,6 +48,12 @@ gen-sdk-dev: dev-env-check-binaries ## Generates a dev version of the Foundation
 		--config ./config/foundation_sdk.dev.yaml \
 		--parameters kind_registry_version=next,grafana_version=main
 
+.PHONY: gen-tests
+gen-tests: dev-env-check-binaries ## Generates the code described by tests schemas.
+	rm -rf generated
+	$(RUN_DEVBOX) go run ./cmd/cli/ generate \
+		--config ./config/foundation_sdk.tests.yaml
```
Comment (spinillos) [inline Makefile:55]: Shouldn't have any reference to the equality tests? Looks like that it executes all tests but they are only for equality ones.

**Review by spinillos — APPROVED** [review]
Looks good. Only a few comments :)

File: Makefile (hunk 4 above)
Comment (K-Phoen) [reply Makefile:55]: This doesn't execute any test, it generates types that some tests depend on (just the equality tests for now, but we'll probably add more in the future)

**Pull request merged**